plugins/trace_api_plugin/store_provider.cpp

```diff
@@ -0,0 +1,83 @@
+#include <eosio/trace_api_plugin/store_provider.hpp>
+
+#include <fc/variant_object.hpp>
+#include <fc/crypto/base64.hpp>
+
+namespace eosio::trace_api_plugin {
+   namespace bfs = boost::filesystem;
+   store_provider::store_provider(const bfs::path& slice_dir, uint32_t width)
+   : _slice_provider(slice_dir, width) {
+   }
+
+   void store_provider::append(const block_trace_v0& bt) {
+      fc::cfile trace;
+      fc::cfile index;
+      find_slice_pair(bt.number, true, trace, index);
+      const uint64_t offset = append_store(bt, trace);
+      auto be = metadata_log_entry { block_entry_v0 { .id = bt.id, .number = bt.number, .offset = offset }};
+      append_store(be, index);
+   }
+
+   void store_provider::append_lib(uint32_t lib) {
+      fc::cfile index;
+      const uint32_t slice_number = _slice_provider.slice_number(lib);
+      _slice_provider.find_index_slice(slice_number, true, index);
+      auto le = metadata_log_entry { lib_entry_v0 { .lib = lib }};
+      append_store(le, index);
+   }
+
+   void store_provider::find_slice_pair(uint32_t block_height, bool append, fc::cfile& trace, fc::cfile& index) {
+      const uint32_t slice_number = _slice_provider.slice_number(block_height);
+      const bool trace_created = _slice_provider.find_trace_slice(slice_number, append, trace);
+      const bool index_created = _slice_provider.find_index_slice(slice_number, append, index);
+      if (trace_created != index_created) {
+         const std::string trace_status = trace_created ? "new" : "existing";
+         const std::string index_status = index_created ? "new" : "existing";
+         elog("Trace file is ${ts}, but it's metadata file is ${is}. This means the files are not consistent.", ("ts", trace_status)("is", index_status));
```

Comment:
More of an integration note for later.   In the extraction and response system we have been feeding in a `std::function<void(const std::exception_ptr&)>` (or if we adopt it a `std::function<void(const exception_with_context&)>`) and letting the plugin code decide what it wants to log and at what level. 

For consistency we will want to do that here as well.  The intent is to not have to worry about custom `fc::logger` s, log levels etc in code this deep

Reply:
This sounds like a plugin level design that we need, so treating this like a next step (i.e. after PR is merged)